### PR TITLE
Improve EOCRecorder docs and checks

### PR DESCRIPTION
**Pull request by alexfikl**
* Adds some types
* Adds methods to docs
* Adds some asserts in `add_data_point` to make sure that it doesn't get weird non-numeric data

### CODE REVIEW

File: pytools/convergence.py [hunk 1]
```diff
@@ -20,13 +22,32 @@ def estimate_order_of_convergence(abscissae, errors):
...
     def __init__(self):
-        self.history = []
+        self.history: List[Tuple[float, float]] = []
+
+    def add_data_point(self, abscissa: float, error: float) -> None:
+        from numbers import Number
+        assert isinstance(abscissa, Number), \
+                f"'abscissa' is not a number: {type(abscissa).__name__}"
+        assert isinstance(error, Number), \
+                f"'error' is not a number: {type(error).__name__}"
```
Comment (inducer) [inline pytools/convergence.py:44]: Input checking should be `if` not `assert`.

**Review by inducer — COMMENTED** [review]
Thanks! LGTM other than the `assert` thing.

File: pytools/convergence.py (hunk 1 above)
Comment (alexfikl) [reply pytools/convergence.py]: Done in new version 3bd3f56.

**Comment by inducer** [comment]
Thx!

**Comment by inducer** [comment]
This may have been a bit overzealous. It extensively breaks meshmode's tests: https://github.com/inducer/pyopencl/runs/4058188593?check_suite_focus=true

**Comment by inducer** [comment]
#109

**Pull request merged**